### PR TITLE
Check vote account initialization

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
Epoch stakes only needs to know about properly sized, initialized Vote accounts

#### Summary of Changes
- Check vote account size on initialization
- Check that vote account is initialized -- ie. non-empty and the correct size -- before it is included in cache. The size check should be redundant, given the vote-processor change.

### CODE REVIEW

File: programs/vote/src/vote_state/mod.rs [hunk 1]
```diff
@@ -569,6 +572,12 @@ impl VoteState {
         self.last_timestamp = BlockTimestamp { slot, timestamp };
         Ok(())
     }
+
+    pub fn is_uninitialized_no_deser(data: &[u8]) -> bool {
+        const VERSION_OFFSET: usize = 4;
+        &data[VERSION_OFFSET..VERSION_OFFSET + DEFAULT_PRIOR_VOTERS_OFFSET]
```
Comment (jackcmay) [inline programs/vote/src/vote_state/mod.rs:578]: Is there a chance that the data size could be less than the indices here?

File: runtime/src/stakes.rs [hunk 2]
```diff
@@ -116,16 +117,19 @@ impl Stakes {
...
             // delegated vote pubkey)
             let old = self.vote_accounts.remove(pubkey);
-            // when account is removed (lamports == 0), don't readd so that given
-            // `pubkey` can be used for any owner in the future, while not
-            // affecting Stakes.
-            if account.lamports != 0 {
+            // when account is removed (lamports == 0 or data uninitialized), don't read so that
+            // given `pubkey` can be used for any owner in the future, while not affecting Stakes.
+            if account.lamports != 0
+                && !(check_vote_init && (account.data.len() != VoteState::size_of())
```
Comment (jackcmay) [inline runtime/src/stakes.rs:130]: Maybe move this size check into `is_unitialized_no_deser`?

File: programs/vote/src/vote_state/mod.rs (hunk 1 above)
Comment (CriesofCarrots) [reply programs/vote/src/vote_state/mod.rs]: Not with the size check coming first ;)

File: runtime/src/stakes.rs (hunk 2 above)
Comment (CriesofCarrots) [reply runtime/src/stakes.rs:130]: Good idea; makes the indexing more clear

Comment (CriesofCarrots) [reply runtime/src/stakes.rs:130]: Adding a unit test for `is_unitialized_no_deser` also

Comment (CriesofCarrots) [reply runtime/src/stakes.rs]: Updated

**Review by jackcmay — APPROVED** [review]

**Pull request merged**